Technical review: Document anchored container queries (#42710)
mdn/content: files/en-us/web/css/guides/anchor_positioning/anchored_container_queries/index.md

---
title: Using anchored container queries
short-title: Anchored container queries
slug: Web/CSS/Guides/Anchor_positioning/Anchored_container_queries
page-type: guide
sidebar: cssref
---

[CSS anchor positioning](/en-US/docs/Web/CSS/Guides/Anchor_positioning) includes mechanisms for providing [fallback options](/en-US/docs/Web/CSS/Guides/Anchor_positioning/Try_options_hiding). These are alternative fallback positions that the browser can try placing an anchor-positioned element in, relative to its anchor, to put it back on-screen if the positioned element starts to overflow the viewport.

> An additional requirement is styling the anchor-positioned element differently depending on which fallback position it is placed in, which is achieved using **anchored container queries**. This guide show how to use anchored container queries, and provides a couple of examples.

Reviewer: ```suggestion
An additional requirement is styling the anchor-positioned element differently depending on which fallback position it is placed in, which is achieved using **anchored container queries**. This guide shows how to use anchored container queries, and provides a couple of examples.
```

- The {{cssxref("@container")}} at-rule `anchored` keyword: This is followed by a set of parentheses inside which the `fallback` descriptor is included. The descriptor's value is a `position-try-fallbacks` value.

For example, let's say we have a tooltip element that is positioned above its anchor by default via a {{cssxref("position-area")}} value of `top`, but has a {{cssxref("position-try-fallbacks")}} value of `flip-block` specified. This will cause the tooltip to flip in the block direction to the bottom of its anchor when it starts to overflow the top of the viewport. If we want to detect when the fallback is applied to the tooltip, we first need to set `container-type: anchored` on it to turn it into an anchored query container.

```css
.tooltip {
  position: absolute;
  position-anchor: --myAnchor;
  position-area: top;
  position-try-fallbacks: flip-block;
  container-type: anchored;
}
```

With this in place, we can now write a container query like so:

```css
@container anchored(fallback: flip-block) {

>   /* Descendent styles here */

Reviewer: ```suggestion
  /* Descendant styles here */
```

}
```

The query test — `anchored(fallback: flip-block)` — will return true when the `flip-block` fallback option is applied to the tooltip, in which case the styles specified within the `@container` block will be applied. You might for example want to change the position and orientation of an arrow so that it points upwards rather than downwards, or change the direction of a gradient.

> [!NOTE]

> > Bear in mind that, as with all container queries, the applied styles can only affect descendents of the container, not the container itself. This might require you to apply some of your positioned element styles to a wrapper element inside it, rather than to the element itself, as demonstrated in [Multiple fallbacks example](#multiple_fallbacks_example).

Reviewer: ```suggestion
> Bear in mind that, as with all container queries, the applied styles can only affect descendants of the container, not the container itself. This might require you to apply some of your positioned element styles to a wrapper element inside it, rather than to the element itself, as demonstrated in [Multiple fallbacks example](#multiple_fallbacks_example).
```

## Basic usage example

> This example includes an anchor element that has a infobox positioned relative to it. Initially, the infobox is positioned above the anchor and includes an arrow pointing down towards the anchor. We include a position try fallback so that the infobox moves below the anchor when the content scrolls up enough that the infobox starts to scroll off the top of the viewport. In addition, we use an anchored container query to change styles once the fallback kicks in, moving the arrow and pointing it upwards instead.

Reviewer: Typo. 

```suggestion
This example includes an anchor element that has an infobox positioned relative to it.
Initially, the infobox is positioned above the anchor and includes an arrow pointing down towards the anchor. We include a position try fallback so that the infobox moves below the anchor when the content scrolls up enough that the infobox starts to scroll off the top of the viewport. In addition, we use an anchored container query to change styles once the fallback kicks in, moving the arrow and pointing it upwards instead.
```

  background-color: black;
  font-size: 1.4em;
  padding: 10px;
  margin: 1rem;
  border-radius: 10px;
}

.infobox::before {
  color: black;
  font-size: 1rem;
  margin: 0;
  line-height: 0.5;
  left: 0;
  width: 100%;
  text-align: center;
}

@supports not (container-type: anchored) {
  body::before {
    content: "Your browser does not support anchored container queries.";
    color: black;
    background-color: #ffcd33;
    display: block;
    width: 100%;
    text-align: center;
    padding: 1rem 0;
  }

  body {
    width: 100%;
    max-width: 100%;
    margin: 0;
  }

  body > * {
    display: none;
  }
}
```

```css live-sample___basic-example
.anchor {
  anchor-name: --myAnchor;
}
```

Next, we give the `infobox` `<div>` a {{cssxref("position")}} value of `fixed` and a {{cssxref("position-anchor")}} value of `--my-anchor` to associate it with the anchor element. We then give the infobox a {{cssxref("position-area")}} value of `top` to position it above the anchor element and a {{cssxref("position-try-fallbacks")}} value of `bottom` so that the infobox will be moved below the anchor when it starts to overflow the top of the viewport as the content is scrolled upwards.

> Finally, we set a {{cssxref("container-type")}} value of `anchored` on the infobox to designate it as an anchored query container, meaning that we can now detect when different `position-try-fallbacks` are active on the infobox via {{cssxref("@container")}} at-rules, and update styles on its descendents as a result.

Reviewer: ```suggestion
Finally, we set a {{cssxref("container-type")}} value of `anchored` on the infobox to designate it as an anchored query container, meaning that we can now detect when different `position-try-fallbacks` are active on the infobox via {{cssxref("@container")}} at-rules, and update styles on its descendants as a result.
```

```css live-sample___basic-example
.infobox {
  position: fixed;
  position-anchor: --myAnchor;
  position-area: top;
  position-try-fallbacks: bottom;
  container-type: anchored;
}
```

Now we'll add the arrow to the infobox using generated content on its {{cssxref("::before")}} pseudo-element. We set the pseudo-element's {{cssxref("content")}} property to a suitable down arrow icon, position it absolutely, and set its {{cssxref("top")}} property to `105%` to position it at the bottom of the infobox (we set it to more than `100%` so that it visually matches up with the position of the corresponding up arrow).

```css live-sample___basic-example
.infobox::before {
  content: "▼";
  position: absolute;
  top: 105%;
}
```

> Now onto the anchored container query. We include a `@container` at-rule with its test defined as `anchored(fallback: bottom)`. This means that when the `bottom` position-try fallback is applied to the inbox, the CSS inside the at-rule is applied to the document. Inside, we define alternative styling for the infobox `::before` pseudo-element that swaps out the down arrow icon for an up arrow and positions it at the top of the infobox.

Reviewer: ```suggestion
Now onto the anchored container query. We include a `@container` at-rule with its test defined as `anchored(fallback: bottom)`. This means that when the `bottom` position-try fallback is applied to the infobox, the CSS inside the at-rule is applied to the document. Inside, we define alternative styling for the infobox `::before` pseudo-element that swaps out the down arrow icon for an up arrow and positions it at the top of the infobox.
```

```css live-sample___basic-example
@container anchored(fallback: bottom) {
  .infobox::before {
    content: "▲";
    bottom: 100%;
    top: auto;
  }
}
```

> [!NOTE]
> There is more CSS included in this example to handle the basic styling of all the elements, but we've only shown you the parts relevant to anchored container queries. To see the full code, open the example in the MDN Playground by pressing the "Play" button on one of the code blocks or the live rendering.

This example renders like so:

{{ EmbedLiveSample("basic-example", "100%", "350") }}

Try scrolling the demo so that the anchor moves near to the top of the viewport and note how, not only does the infobox move below the anchor to remain on the screen, but the styling also updates so that the arrow icon still works for the new infobox position.

If you scroll the anchor back down towards the bottom of the viewport, the infobox will move back up above it again.

## Multiple fallbacks example

> This example shows multiple position-try fallbacks and anchored container queries in action, and also addresses the problem of what to do if you want to use anchored container queries to set styles on the anchor-positioned element itself, rather than its descendents, using an extra wrapper element. The example also includes some JavaScript that allows you to move the anchor element around the screen using the mouse or the keyboard to check out the different fallbacks.

Reviewer: ```suggestion
This example shows multiple position-try fallbacks and anchored container queries in action, and also addresses the problem of what to do if you want to use anchored container queries to set styles on the anchor-positioned element itself, rather than its descendants, using an extra wrapper element. The example also includes some JavaScript that allows you to move the anchor element around the screen using the mouse or the keyboard to check out the different fallbacks.
```

Finally, we set a {{cssxref("container-type")}} value of `anchored` on the infobox to designate it as an anchored query container, meaning that we can now detect when different `position-try-fallbacks` are active on the infobox via {{cssxref("@container")}} at-rules, and update styles on its descendants as a result.

```css live-sample___basic-example
.infobox {
  position: fixed;
  position-anchor: --myAnchor;
  position-area: top;
  position-try-fallbacks: bottom;
  container-type: anchored;
}
```

Now we'll add the arrow to the infobox using generated content on its {{cssxref("::before")}} pseudo-element. We set the pseudo-element's {{cssxref("content")}} property to a suitable down arrow icon, position it absolutely, and set its {{cssxref("top")}} property to `105%` to position it at the bottom of the infobox (we set it to more than `100%` so that it visually matches up with the position of the corresponding up arrow).

```css live-sample___basic-example
.infobox::before {
  content: "▼";
  position: absolute;
  top: 105%;
}
```

> Now onto the anchored container query. We include a `@container` at-rule with its test defined as `anchored(fallback: bottom)`. This means that when the `bottom` position-try fallback is applied to the infobox, the CSS inside the at-rule is applied to the document. Inside, we define alternative styling for the infobox `::before` pseudo-element that swaps out the down arrow icon for an up arrow and positions it at the top of the infobox.

Reviewer: Maybe avoid too many Now in a row.

```suggestion
Next we add the anchored container query. We include a `@container` at-rule with its test defined as `anchored(fallback: bottom)`. This means that when the `bottom` position-try fallback is applied to the infobox, the CSS inside the at-rule is applied to the document. Inside, we define alternative styling for the infobox `::before` pseudo-element that swaps out the down arrow icon for an up arrow and positions it at the top of the infobox.
```

```css live-sample___basic-example
@container anchored(fallback: bottom) {
  .infobox::before {
    content: "▲";
    bottom: 100%;
    top: auto;
  }
}
```

> [!NOTE]
> There is more CSS included in this example to handle the basic styling of all the elements, but we've only shown you the parts relevant to anchored container queries. To see the full code, open the example in the MDN Playground by pressing the "Play" button on one of the code blocks or the live rendering.

This example renders like so:

{{ EmbedLiveSample("basic-example", "100%", "350") }}

> Try scrolling the demo so that the anchor moves near to the top of the viewport and note how, not only does the infobox move below the anchor to remain on the screen, but the styling also updates so that the arrow icon still works for the new infobox position.

Reviewer: This is very cool.
Note, I was surprised that scrolling worked well both when scrolling up and down - i.e. it flips both ways.
I guess I don't understand how the `bottom` positioning and fallback works. I.e. when you scroll down the infobox is about to overflow so the fallback of bottom triggers. But at this point it is already at the bottom, so how does it end up at the top. A mystery, but _perhaps_ not one for you to cover here.
